### PR TITLE
Added FractionInputInteractionViewTest.kt

### LINKED ISSUE #4135: Add tests for FractionInputInteractionView
FractionInputInteractionView should have dedicated tests to verify that it behaves correctly as an interaction. Today, it's tested through InputInteractionViewTestActivity, but these tests ought to be moved to a dedicated test suite for fraction input, and other fraction-specific tests should be added.

### PR DESCRIPTION
**Pull request by bhaktideshmukh**
## Explanation
Fixes #4135 : Added FractionInputInteractionViewTest.kt

## Essential Checklist

- [ ] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [ ] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [ ] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [ ] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

### CODE REVIEW

File: app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+  }
+
+  @Test
+  fun testFractionInput_withNoInput_hasCorrectPendingAnswerType() {
+    val activityScenario = ActivityScenario.launch(
+      InputInteractionViewTestActivity::class.java
+    )
+    activityScenario.onActivity { activity ->
+      val pendingAnswer = activity.fractionInteractionViewModel.getPendingAnswer()
+      Truth.assertThat(pendingAnswer.answer).isInstanceOf(InteractionObject::class.java)
```
Comment (rt4914) [inline app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt]: Please prefer direct imports here and everywhere else.
`assertThat(pendingAnswer.answer).isInstanceOf(InteractionObject::class.java)`

File: app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+    }
+  }
+
+  @Test
+  @DisableAccessibilityChecks
+  fun testFractionInput_withNegativeNumber_hasCorrectPendingAnswer() {
+    val activityScenario = ActivityScenario.launch(
+      InputInteractionViewTestActivity::class.java
+    )
+    Espresso.onView(ViewMatchers.withId(R.id.test_fraction_input_interaction_view))
```
Comment (rt4914) [inline app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt]: Please prefer direct imports here and everywhere else.
`onView(withId(R.id.test_fraction_input_interaction_view))`

**Review by rt4914 — CHANGES_REQUESTED** [review]
@anandwana001 Can you please take a lead on this PR? I am not sure why I am getting assigned as the code-owner. (I will update this later on in a separate PR).

Thanks.

**Comment by anandwana001** [comment]
@bhaktideshmukh 

```
Today, it's tested through InputInteractionViewTestActivity, but these tests ought to be moved to a 
dedicated test suite for fraction input and other fraction-specific tests should be added.
```
After reading the issue description, what I understand is currently we use `InputInteractionViewTestActivity` activity to test  different interactions. `[FractionInputInteractionView], [NumericInputInteractionView],and [TextInputInteractionView]`.
As for this issue, we need to move the fraction part completely to the new test suite.

So, what we will require here is a new `FractionInputInteractionViewTestActivity` which does all the initialization work for the fraction part and all the current fraction tests which are using `InputInteractionViewTestActivity` must use the new fraction specific test activity.

Also, as part of this issue, we need to write new test cases which cover all the edge cases for fractions.
` and other fraction-specific tests should be added.`

Also, check this comment here - https://github.com/oppia/oppia-android/issues/4135#issuecomment-1035921300

@BenHenning Let me know if I misunderstand anything here.

File: app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+      Truth.assertThat(pendingAnswer.answer.fraction.isNegative).isEqualTo(true)
+      Truth.assertThat(pendingAnswer.answer.fraction.wholeNumber).isEqualTo(9)
+    }
+  }
+
+  @Test
+  @DisableAccessibilityChecks
+  fun testFractionInput_withWholeNumber_hasCorrectPendingAnswer() {
+    val activityScenario = ActivityScenario.launch(
+      InputInteractionViewTestActivity::class.java
```
Comment (anandwana001) [inline app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt]: Replace with the new Fraction specific test activity.

File: app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+import javax.inject.Singleton
+
+/** Tests for [FractionInputInteractionView]. */
+@RunWith(AndroidJUnit4::class)
+@Config(
+  application = FractionInputInteractionViewTest.TestApplication::class,
+  qualifiers = "port-xxhdpi"
+)
+@LooperMode(LooperMode.Mode.PAUSED)
+
```
Comment (anandwana001) [inline app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt]:
```suggestion
```

**Review by anandwana001 — CHANGES_REQUESTED** [review]
Left nit thought and a comment above. 

1. Update `InputInteractionViewTestActivity` to remove all the fraction part. 
2. Add test result screenshots.

File: app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt (hunk 3 above)
Comment (bhaktideshmukh) [reply app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt]: > Replace with the new Fraction specific test activity.

I tried the way you suggested, by creating FractionInputInteractionViewTestActivity to replace InputInteractionViewTestActivity but now  tests are failing in FractionInputInteractionViewTest.kt file.

Screenshot from 2022-03-19 00-30-43

Specific error:--
'java.lang.RuntimeException: Unable to resolve activity for Intent { act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] cmp=org.oppia.android/.app.testing.FractionInputInteractionViewTestActivity } -- see https://github.com/robolectric/robolectric/pull/4736 for details'

Please do suggest how should I proceed?

Comment (anandwana001) [reply app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt]: Could you share your `FractionInputInteractionViewTestActivity`?

Also, did you add it in the manifest file? like `<activity android:name=".app.testing.InputInteractionViewTestActivity" />`

Comment (bhaktideshmukh) [reply app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt]: I have pushed the recent changes.

File: app/src/main/java/org/oppia/android/app/testing/FractionInputInteractionViewTestActivity.kt [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import org.oppia.android.app.model.Interaction
+import org.oppia.android.app.model.WrittenTranslationContext
+import org.oppia.android.app.player.state.answerhandling.AnswerErrorCategory
+import org.oppia.android.app.player.state.answerhandling.InteractionAnswerErrorOrAvailabilityCheckReceiver
+import org.oppia.android.app.player.state.itemviewmodel.FractionInteractionViewModel
+import org.oppia.android.app.player.state.listener.StateKeyboardButtonListener
+import org.oppia.android.app.translation.AppLanguageResourceHandler
+import org.oppia.android.domain.translation.TranslationController
+import javax.inject.Inject
+
```
Comment (anandwana001) [inline app/src/main/java/org/oppia/android/app/testing/FractionInputInteractionViewTestActivity.kt:19]: Add comment.

```
/**
 * This is a dummy activity to test input interaction views.
 * It contains [FractionInputInteractionView].
 */
```

File: app/src/main/java/org/oppia/android/app/testing/FractionInputInteractionViewTestActivity.kt [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  StateKeyboardButtonListener,
+  InteractionAnswerErrorOrAvailabilityCheckReceiver {
+
+  @Inject
+  lateinit var resourceHandler: AppLanguageResourceHandler
+
+  @Inject
+  lateinit var translationController: TranslationController
+
+  val fractionInteractionViewModel by lazy {
```
Comment (anandwana001) [inline app/src/main/java/org/oppia/android/app/testing/FractionInputInteractionViewTestActivity.kt]: This class is missing `onCreate` function.
Take a look at the similar function in the `InputInteractionViewTestActivity`.  Add the UI logic too. 

Also, see `activity_input_interaction_view_test`, how we are binding the UI with the `InputInteractionViewTestActivity`. Now, for this PR, we need to separate out the fraction part completely in the new UI file as well.

File: app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+class FractionInputInteractionViewTest {
```
Comment (anandwana001) [inline app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt:101]: Run these tests on espresso and add a screenshot in the PR description showing that all the tests are passing.

File: app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt (hunk 4 above)
Comment (anandwana001) [reply app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt]: remove the empty line above.

File: app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+      }
+      onView(withId(R.id.test_fraction_input_interaction_view)).check(matches(isDisplayed()))
+        .check(matches(withText("9/5")))
+    }
+  }
+
+  @Singleton
+  @Component(
+    modules = [
+      StateFragmentTest.TestModule::class, RobolectricModule::class, PlatformParameterModule::class,
```
Comment (anandwana001) [inline app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt:524]: Do we need this? `StateFragmentTest.TestModule`

File: app/src/main/AndroidManifest.xml [hunk 9]
```diff
@@ -271,6 +271,9 @@
     <activity
       android:name=".app.testing.activity.TestActivity"
       android:theme="@style/OppiaThemeWithoutActionBar" />
+    <activity
+      android:name=".app.testing.FractionInputInteractionViewTestActivity"
+      android:theme="@style/OppiaThemeWithoutActionBar" />
```
Comment (anandwana001) [inline app/src/main/AndroidManifest.xml]:
```xml
<activity android:name=".app.testing.FractionInputInteractionViewTestActivity" />
```

Comment (anandwana001) [reply app/src/main/AndroidManifest.xml]: Also, keep it below the line -

```xml
<activity android:name=".app.testing.InputInteractionViewTestActivity" />
```
Good for readability.

File: app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt (hunk 7 above)
Comment (anandwana001) [inline app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt:101]:
```suggestion
class FractionInputInteractionViewTestActivityTest {
```

File: app/src/main/java/org/oppia/android/app/testing/FractionInputInteractionViewTestActivity.kt [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+import org.oppia.android.domain.translation.TranslationController
+import javax.inject.Inject
+
+class FractionInputInteractionViewTestActivity :
+  InjectableAppCompatActivity(),
+  StateKeyboardButtonListener,
+  InteractionAnswerErrorOrAvailabilityCheckReceiver {
+
+  @Inject
+  lateinit var resourceHandler: AppLanguageResourceHandler
```
Comment (anandwana001) [inline app/src/main/java/org/oppia/android/app/testing/FractionInputInteractionViewTestActivity.kt:28]: We had introduced inject function for this file in `ActivityComponentImp`, but we are not calling it. 

The way we call the inject function is 

```kt
(activityComponent as ActivityComponentImpl).inject(this)
```
without this, these `@inject` dependency injections will not work.

**Review by anandwana001 — CHANGES_REQUESTED** [review]
Update the `InputInteractionViewTestActivity` as well by removing the logic of `FractionInteractionViewModel` as this PR is all about separating out the fraction from it. 

Left few more suggestions.  

Also, check for failing github actions and update the PR with the latest develop branch.

**Comment by bhaktideshmukh** [comment]
@anandwana001 
I did all the changes the way you suggested but while running the tests I am facing this problem:--
'API level 30 is not available'
how can I fix this?


Screenshot from 2022-03-22 23-07-22

**Comment by anandwana001** [comment]
Is you PR updated with the latest develop branch? If not let's try update and re run.

**Comment by bhaktideshmukh** [comment]
> Is you PR updated with the latest develop branch? If not let's try update and re run.

Its still the same

File: app/src/main/java/org/oppia/android/app/testing/FractionInputInteractionViewTestActivity.kt (hunk 5 above)
Comment (bhaktideshmukh) [reply app/src/main/java/org/oppia/android/app/testing/FractionInputInteractionViewTestActivity.kt:19]: done

File: app/src/main/java/org/oppia/android/app/testing/FractionInputInteractionViewTestActivity.kt (hunk 6 above)
Comment (bhaktideshmukh) [reply app/src/main/java/org/oppia/android/app/testing/FractionInputInteractionViewTestActivity.kt]: done

File: app/src/main/AndroidManifest.xml (hunk 9 above)
Comment (bhaktideshmukh) [reply app/src/main/AndroidManifest.xml]: done

File: app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt (hunk 7 above)
Comment (bhaktideshmukh) [reply app/src/test/java/org/oppia/android/app/customview/interaction/FractionInputInteractionViewTest.kt:101]: > 

Why do we need to change the name?

**Comment by bhaktideshmukh** [comment]
@anandwana001 I guess I have messed this PR So I will release a new and clean PR for this issue with all your previously suggested changes.
